controller/object_store_controller.go

```diff
@@ -0,0 +1,870 @@
+	if err != nil || pvc.Status.Phase != corev1.ClaimBound {
+		osc.setObjectStoreState(store, longhorn.ObjectStoreStateError)
+		return err
+	}
+
+	_, err = osc.ds.GetVolume(genPVName(store))
+	if err != nil {
+		osc.setObjectStoreState(store, longhorn.ObjectStoreStateError)
+		return err
+	}
+
+	_, err = osc.ds.GetPersistentVolume(genPVName(store))
+	if err != nil {
+		osc.setObjectStoreState(store, longhorn.ObjectStoreStateError)
+		return err
+	}
+	return nil
+}
+
+func (osc *ObjectStoreController) handleStopping(store *longhorn.ObjectStore) (err error) {
+	dpl, err := osc.ds.GetDeployment(store.Name)
+	if err != nil {
+	} else if *dpl.Spec.Replicas != 0 {
+		dpl.Spec.Replicas = int32Ptr(0)
+		_, err = osc.ds.UpdateDeployment(dpl)
+		return err
+	} else if dpl.Status.AvailableReplicas > 0 {
+		return nil // wait for shutdown
+	}
+
+	osc.setObjectStoreState(store, longhorn.ObjectStoreStateStopped)
+	return nil
+}
+
+func (osc *ObjectStoreController) handleStopped(store *longhorn.ObjectStore) (err error) {
+	if store.Spec.TargetState == longhorn.ObjectStoreStateRunning {
+		osc.setObjectStoreState(store, longhorn.ObjectStoreStateStarting)
+		return nil
+	}
+	return nil
+}
+
+func (osc *ObjectStoreController) handleTerminating(store *longhorn.ObjectStore) (err error) {
+	// Just remove the finalizer. The K8s API will take care of removing dependent
+	// resources.
+	return osc.ds.RemoveFinalizerForObjectStore(store)
+}
+
+func (osc *ObjectStoreController) setObjectStoreState(
+	store *longhorn.ObjectStore,
+	status longhorn.ObjectStoreState,
+) (*longhorn.ObjectStore, error) {
+	store.Status.State = status
+	return osc.ds.UpdateObjectStoreStatus(store)
+}
+
+func (osc *ObjectStoreController) initializeObjectStore(store *longhorn.ObjectStore) (err error) {
+	// TODO: check if we really need/want to update the spec's image here
+	if store.Spec.Image == "" {
+		store.Spec.Image = osc.s3gwImage
+	}
+
+	if store.Spec.UiImage == "" {
+		store.Spec.UiImage = osc.uiImage
+	}
+
+	store.Spec.TargetState = longhorn.ObjectStoreStateRunning
+
+	store, err = osc.ds.UpdateObjectStore(store)
+	if err != nil {
+		return nil // retry until no more conflicts
+	}
+
+	_, err = osc.setObjectStoreState(store, longhorn.ObjectStoreStateStarting)
+	return err
+}
+
+func (osc *ObjectStoreController) getOrCreatePVC(store *longhorn.ObjectStore) (*corev1.PersistentVolumeClaim, *longhorn.ObjectStore, error) {
+	pvc, err := osc.ds.GetPersistentVolumeClaim(osc.namespace, genPVCName(store))
+	if err == nil {
+		return pvc, store, nil
+	} else if datastore.ErrorIsNotFound(err) {
+		pvc, err = osc.createPVC(store)
+		if err != nil {
+			store, _ = osc.setObjectStoreState(store, longhorn.ObjectStoreStateError)
+			return nil, store, errors.Wrap(err, "failed to create persistent volume claim")
+		} else if store.Status.State != longhorn.ObjectStoreStateStarting {
+			store, _ = osc.setObjectStoreState(store, longhorn.ObjectStoreStateStarting)
```

Comment:
Same as comment above: suggesting to defer the updating the ObjectStore object in a defer function here so that we don't call multiple API update in the middle of the sync loop.